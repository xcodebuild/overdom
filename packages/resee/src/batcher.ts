const tasks: Set<Function> = new Set();
let timer: NodeJS.Timeout | number | null = null;

export function schedule(task: Function) {
  tasks.add(task);
  if (!timer) {
    timer = setTimeout(() => {
      timer = null;
      const taskItems = Array.from(tasks.values());
      tasks.clear();
      taskItems.forEach(taskItem => {
        taskItem();
      });
    }, 0);
  }
}
