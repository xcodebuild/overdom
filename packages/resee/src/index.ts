export { h, render } from './h';

export { reactive, autorun, run } from './reactive';

export { $if, $map } from './directive';
