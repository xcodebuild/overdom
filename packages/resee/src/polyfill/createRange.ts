if (!document.createRange) {
  class Range {
    start?: number;
    end?: number;
    container?: Node;
    afterEndNode?: Node | null;

    detach() {
      // pass
    }

    setStart(container: Node, offset: number) {
      this.start = offset;
      this.container = container;
    }
    setEnd(container: Node, offset: number) {
      this.end = offset;
      this.container = container;
    }
    deleteContents() {
      const nodes = Array.from(this.container!.childNodes);
      const lastNode = nodes[nodes.length - 1];
      this.afterEndNode = lastNode?.nextSibling || null;
      nodes.forEach((node, index) => {
        if (index >= this.start! && index <= this.end!) {
          node.parentNode?.removeChild(node);
        }
      });
    }
    insertNode(node: Node | DocumentFragment) {
      this.container!.insertBefore(node, this.afterEndNode!);
    }
  }
  // @ts-ignore
  document.createRange = function() {
    return new Range();
  };
}
