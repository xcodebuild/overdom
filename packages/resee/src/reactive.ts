const noop = () => {};

export type ReactiveHandler<T> = ((newValue?: T) => T) &
  {
    [K in keyof T]: ReactiveHandler<T[K]>;
  } & {
    __internal_reactive__: Reactive<T>;
    unsubscribe: Reactive<T>['unsubscribe'];
    subscribe: Reactive<T>['subscribe'];
  };

type Key = string | number;

type Tracker = (args: {
  unsubscribe: () => void;
  remapKeys: (keyMap: Record<Key, Key>) => void;
}) => void;

const computedTrackerFns: Tracker[] = [];

type UnReactiveTypeOf<T> = T extends () => infer I ? I : T;

export function unReactive<T>(handler: T) {
  return isReactive(handler)
    ? (((handler as unknown) as ReactiveHandler<
        any
      >).__internal_reactive__.selector() as UnReactiveTypeOf<T>)
    : handler;
}

class Reactive<T extends Object> {
  selector: () => T;
  setter: (val: T) => void;

  handler: ReactiveHandler<T>;
  trackers: Set<Tracker> = new Set();
  childs = new Map<Key, ReactiveHandler<T>>();

  unsubscribe = (tracker: Tracker) => {
    this.trackers.delete(tracker);
  };

  subscribe = (tracker: Tracker) => {
    this.trackers.add(tracker);
  };

  trigger() {
    const trackers = Array.from(this.trackers.values());
    trackers.forEach(tracker =>
      tracker({
        unsubscribe: () => this.unsubscribe(tracker),
        remapKeys: (keyMap: Record<Key, Key>) => {
          Object.keys(keyMap).forEach(key => {
            const nextKey = keyMap[key];
            const child = this.childs.get(key);
            if (child) {
              child.__internal_reactive__.selector = () => {
                // @ts-ignore
                return this.selector()[nextKey];
              };
              this.childs.set(nextKey, child);
              this.childs.delete(key);
            }
          });
        },
      })
    );

    this.childs.forEach(child => child.__internal_reactive__.trigger());
  }

  constructor(baseSelector: () => T, baseSetter: (val: T) => void) {
    this.selector = baseSelector;
    this.setter = baseSetter;

    this.handler = (new Proxy(noop, {
      get: (_, key: string) => {
        if (key === '__internal_reactive__') {
          return this;
        } else if (key === 'subscribe' || key === 'unsubscribe') {
          return this[key];
        }
        if (this.childs.has(key)) {
          return this.childs.get(key);
        }
        // @ts-ignore
        const nextSelector = () => this.selector()[key];
        const nextSetter = (val: any) => {
          this.setter({
            ...this.selector(),
            [key]: val,
          });
        };

        const child = new Reactive(nextSelector, nextSetter).get();
        this.childs.set(key, child);
        return child;
      },
      apply: (_, __, args) => {
        if (args.length === 0) {
          if (computedTrackerFns.length > 0) {
            this.subscribe(computedTrackerFns[computedTrackerFns.length - 1]);
          }
          return this.selector();
        }
        // set
        this.setter(args[0]);
        this.trigger();
        return this.selector();
      },
    }) as unknown) as ReactiveHandler<T>;
  }

  get() {
    return this.handler;
  }
}

export function reactive<T extends Object>(initalState: T) {
  if (isReactive(initalState)) {
    return initalState as unknown as ReactiveHandler<UnReactiveTypeOf<T>>;
  }
  let state = initalState;
  return new Reactive(
    () => {
      return state;
    },
    (newValue: T) => {
      state = newValue;
    }
  ).get() as unknown as ReactiveHandler<UnReactiveTypeOf<T>>;
}

export function isReactive(target: unknown) {
  return target && (target as ReactiveHandler<unknown>).__internal_reactive__
    ? true
    : false;
}

/**
 * Reactive functions
 * const a = reactive(1);
 * Use a function to build a reactive value based on compute of another reactive value
 * const b = reactive(() => a() * 4);
 * @param reactiveFn
 */
export function computed<T>(reactiveFn: () => T) {
  let state: ReactiveHandler<T>;

  const compute = () => {
    computedTrackerFns.push(tracker);
    // first time compute
    const value = reactiveFn();
    computedTrackerFns.pop();
    return value;
  };
  const tracker: Tracker = options => {
    // changed
    // remove old tracker
    options.unsubscribe();

    // recompute
    const value = compute();
    state(value as any);
  };

  const initalValue = compute();
  // @ts-ignore
  state = reactive(initalValue);
  return state;
}
