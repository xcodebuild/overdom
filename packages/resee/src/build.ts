import { computed, ReactiveHandler } from './reactive';

export function buildText(value: ReactiveHandler<string>) {
  const textNode = document.createTextNode(value());
  computed(() => {
    const newValue = value();
    textNode.nodeValue = newValue;
  });
  return textNode;
}
