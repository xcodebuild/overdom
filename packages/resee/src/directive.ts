import { schedule } from './batcher';
import { Fragment, FragmentList } from './fragment';
import { autorun, reactive, Ref, wrapFnHideRefMode } from './reactive';

export function directive<T extends () => Fragment>(func: T) {
  // @ts-ignore
  func.__isdirective__ = true;
  return func;
}

export function isDirective(val: any) {
  return val?.__isdirective__ === true;
}

export function $if(
  cond: () => boolean,
  yes: () => Fragment,
  no: () => Fragment
) {
  return directive(() => {
    let lastResult: boolean;
    let inited = false;
    let lastFragment: Fragment;

    autorun(() => {
      const newResult = cond();
      if (!inited) {
        lastResult = newResult;
        lastFragment = lastResult! ? yes() : no();
        inited = true;
        return;
      }

      const newFragment = newResult ? yes() : no();
      schedule(() => {
        lastFragment.replaceWith(newFragment);
        lastFragment = newFragment;
      });
      lastResult = newResult;
    });

    return lastFragment!;
  });
}

function createReactiveItem<T>(value: T, index: number) {
  return reactive({
    value,
    index,
  });
}

export type ReactiveItem<T> = {
  value: T;
  index: number;
};

export function $map<T>(
  rawList: T[],
  mapFunc: (item: ReactiveItem<T>) => Fragment,
  keyFunc: (item: ReactiveItem<T>) => string | number
) {
  const list = (rawList as any) as Ref;
  const listData = (list as any) as T[];

  let lastReactiveItems = listData.map((item, index) => {
    return createReactiveItem(item, index);
  });

  const fragmentList = new FragmentList();

  const mapFuncWrap = wrapFnHideRefMode((item: ReactiveItem<T>) => {
    const fragment = mapFunc(item);
    fragment.reactiveItem = item;
    return fragment;
  });

  const keyFuncWrap = wrapFnHideRefMode((item: ReactiveItem<T>) => {
    const r = keyFunc(item);
    if (typeof r !== 'string' && typeof r !== 'number') {
      console.trace(
        'key of $map should return string or number, but got ' + r,
        keyFunc.toString()
      );
    }
    return r;
  });

  let lastFragments = listData.map((_, index) =>
    mapFuncWrap(lastReactiveItems[index])
  );
  let lastKeys = listData.map((_, index) =>
    keyFuncWrap(lastReactiveItems[index])
  );

  lastFragments.forEach((frag, index) => {
    fragmentList.appendFragment(frag, lastKeys[index]);
  });

  let inited = false;

  autorun(notrack => {
    if (!inited) {
      // @ts-ignore
      inited = list.length;
      inited = true;
      return;
    }

    const newList = (list as unknown) as any[];

    notrack(() => {
      const newKeys = newList.map((item, index) =>
        keyFuncWrap(createReactiveItem(item, index))
      );

      const lastKeys = fragmentList.childFragments.map(item => item.key!);

      // diff and patch from keys
      newKeys.forEach((key, index) => {
        const lastKeys = fragmentList.childFragments.map(item => item.key!);
        const lastIndex = lastKeys.indexOf(key);
        if (lastIndex === -1) {
          // insert
          // can not find in old list, mean insert
          // insert to current index
          const reactiveItem = createReactiveItem(newList[index], index);

          const fragment = mapFuncWrap(reactiveItem);

          fragmentList.insert(index, fragment, key);
        } else if (lastIndex !== index) {
          // move

          const fromIndex = lastIndex;
          const toIndex = index;

          fragmentList.move(fromIndex, toIndex);
        }
      });

      const removed = lastKeys.filter(key => newKeys.indexOf(key) === -1);
      fragmentList.removeWithKeys(removed);

      // apply new reactiveIndex
      fragmentList.childFragments.forEach(
        (child, index) => (child!.reactiveItem!.index = index)
      );
    });
  });

  return fragmentList;
}
