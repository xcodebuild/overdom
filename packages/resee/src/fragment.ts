import { schedule } from './batcher';
import { ReactiveItem } from './directive';
import { insert } from './utils';

export class Fragment {
  fragment = document.createDocumentFragment();
  container?: Node;
  childNodes: Node[] = [];
  key?: number | string;
  reactiveItem?: ReactiveItem<any>;

  static isFragment(obj: unknown) {
    if (!obj) {
      return;
    }
    const constructor = (obj as Object).constructor;
    return constructor === Fragment || constructor === FragmentList;
  }

  appendChild(child: Node) {
    const node = child as Node;
    this.fragment.appendChild(node);
    this.childNodes.push(node);
  }

  appendToContainer(container: Node) {
    this.container = container;
    this.createEmpty();
    container.appendChild(this.fragment);
  }

  replaceWith(fragment: Fragment) {
    fragment.container = this.container;
    const range = this.createRange();
    range.deleteContents();
    range.insertNode(fragment.fragment);
    range.detach();
  }

  createEmpty() {
    if (this.childNodes.length === 0) {
      this.appendChild(document.createComment('empty'));
    }
  }

  insertBeforeToContainer(container: Node, beforeNode: Node | null) {
    if (this.container) {
      // has a container already
      // remove self from current container first
      schedule(() => {
        this.fragment = document.createDocumentFragment();
        this.childNodes.forEach(node => {
          this.container!.removeChild(node);
          this.fragment.appendChild(node);
        });
      });
    }
    this.container = container;
    this.createEmpty();
    schedule(() => {
      container.insertBefore(this.fragment, beforeNode);
    });
  }

  createRange() {
    const parent = this.childNodes[0].parentNode!;
    const range = document.createRange();
    const startPos = Array.prototype.indexOf.call(
      parent?.childNodes,
      this.childNodes[0]
    );
    range.setStart(parent, startPos);
    range.setEnd(parent, startPos + this.childNodes.length);
    return range;
  }

  getFirstNode() {
    return this.childNodes[0];
  }
}

export class FragmentList extends Fragment {
  childFragments: Fragment[] = [];

  appendFragment(child: Fragment, key: string | number) {
    child.appendToContainer(this.fragment);
    this.childFragments.push(child);
    child.key = key;
  }

  insert(index: number, fragment: Fragment, newKey: string | number) {
    fragment.key = newKey;
    fragment.insertBeforeToContainer(
      this.container!,
      this.childFragments[index]?.getFirstNode() || null
    );
    insert(this.childFragments, index, fragment);
  }

  removeWithKeys(keys: (number | string)[]) {
    this.childFragments.forEach((child, index) => {
      if (keys.indexOf(child.key!) === -1) {
        return;
      }
      const fragment = child;
      schedule(() => {
        const range = fragment.createRange();
        range.deleteContents();
        range.detach();
      });
      this.childFragments.splice(index, 1);
    });
  }

  move(fromIndex: number, toIndex: number) {
    const insertBefore = this.childFragments[toIndex]

    let insertBeforeNode: Node | null = insertBefore?.getFirstNode();

    const target = this.childFragments[fromIndex]!;
    target.insertBeforeToContainer(
      this.container!,
      insertBeforeNode,
    );

    insert(this.childFragments, toIndex, target);
    let deleteIndex = fromIndex;
    if (toIndex < fromIndex) {
      deleteIndex += 1;
    }
    this.childFragments.splice(deleteIndex, 1);
  }
}
