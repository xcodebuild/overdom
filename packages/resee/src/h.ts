import { schedule } from './batcher';
import { isDirective } from './directive';
import { isReactive, reactive, ReactiveHandler } from './reactive';
import { Fragment } from './fragment';

function buildComponent(
  comp: (props?: Record<string, any>) => Fragment,
  props: Record<string, any> = {}
) {
  const fragment = comp(props);
  return fragment;
}

export type Component<P = undefined> = (props: {
  [K in keyof P]: ReactiveHandler<P[K]>
}) => Fragment;

export function component<P extends Record<string, any>>(func: (props: {
  [K in keyof P]: ReactiveHandler<P[K]>
}) => Fragment) {

  return (props: P) => {
    // @ts-ignore
    const reactiveProps: {
      [K in keyof P]: ReactiveHandler<P[K]>
    } = {};
    if (props) {
      Object.keys(props).forEach(key => {
        // @ts-ignore
        reactiveProps[key] = reactive(props[key]);
      });
    }
    return func(reactiveProps);
  }
}

function setAttr(node: HTMLElement, key: string, value: any) {
  if (key === 'value') {
    // @ts-ignore
    node.value = value;
    return;
  } else if (key === 'class') {
    console.error('should be className, but got class');
    return;
  } else if (key === 'className') {
    key = 'class';
  } else if (key === 'dangerouslySetInnerHTML') {
    const html = value.__html;
    node.innerHTML = html;
    return;
  }
  node.setAttribute(key, value);
}

function bindAttr(node: HTMLElement, key: string, value: ReactiveHandler<any>) {
  let lastValue = value();
  schedule(() => {
    setAttr(node, key, lastValue);
  });

  value.subscribe(() => {
    const newValue = value();
    if (newValue !== lastValue) {
      schedule(() => {
        setAttr(node, key, newValue);
      });
      lastValue = newValue;
    }
  });
}

export function h(
  type: string | Function,
  props: Record<string, string>,
  ...children: string[]
) {
  const fragment = new Fragment();
  if (typeof type === 'string') {
    const tag = document.createElement(type);
    if (props) {
      Object.keys(props).forEach(key => {
        const value = props[key];
        if (typeof value === 'function') {
          if (/^on[A-Z]/.test(key)) {
            tag.addEventListener(
              key.replace(/^on/, '').toLocaleLowerCase(),
              value
            );
            return;
          } else if (isReactive(value)) {
            bindAttr(tag, key, value);
            return;
          }
        }
        setAttr(tag, key, value);
      });
    }
    if (children) {
      children.forEach(child => {
        if (typeof child === 'string') {
          tag.appendChild(document.createTextNode(child));
        } else if (isReactive(child)) {
          const reactiveVal = child as ReactiveHandler<any>;
          const textNode = document.createTextNode(reactiveVal());
          reactiveVal.__internal_reactive__.subscribe(() => {
            schedule(() => {
              textNode.nodeValue = reactiveVal();
            });
          });
          tag.appendChild(textNode);
        } else if (isDirective(child)) {
          const directive = child as () => Fragment;
          const fragment = directive();
          fragment.appendToContainer(tag);
        } else if (Fragment.isFragment(child)) {
          (child as Fragment).appendToContainer(tag);
        } else {
          tag.appendChild(child);
        }
      });
    }
    fragment.appendChild(tag);
  } else if (typeof type === 'function') {
    // component
   
    return buildComponent(type as () => Fragment, {
      ...props,
      children: children,
    });
  }
  return fragment;
}

export function render(fragment: Fragment, container: HTMLElement) {
  fragment.appendToContainer(container);
}

declare global {
  namespace JSX {
    interface IntrinsicElements {
      [name: string]: any;
    }
  }
}
