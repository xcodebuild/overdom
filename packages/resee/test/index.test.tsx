import { render, h, reactive, $if } from '../src';
import { schedule } from '../src/batcher';
import { $map } from '../src/directive';
import { component } from '../src/h';
import '../src/polyfill/createRange';

const waitBatch = () =>
  new Promise(r => {
    schedule(r);
  });

describe('Basic', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
  });
  it('should render hello world', async () => {
    render(<div>Hello World</div>, document.body);
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can reactive count', async () => {
    const App = () => {
      const count = reactive(0);
      const handleClick = () => {
        count(count() + 1);
      };
      return <button onClick={handleClick}>Count {count}</button>;
    };
    render(<App />, document.body);
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

    document.body.querySelector('button')!.click();
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can render mutiple items', async () => {
    const App = () => {
      const count = reactive(0);
      const handleClick = () => {
        count(count() + 1);
      };
      return (
        <div>
          <button onClick={handleClick}>CLICK</button>
          Count {count}
        </div>
      );
    };
    render(<App />, document.body);
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

    document.body.querySelector('button')!.click();
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('$if works', async () => {
    const App = () => {
      const count = reactive(1);

      const handleClick = () => {
        count(count() + 1);
      };

      return (
        <div>
          <button onClick={handleClick}>Count {count}</button>
          <div>
            Bigger than 2
            {$if(
              () => count() > 2,
              () => (
                <span>YES {count}</span>
              ),
              () => (
                <span>NO {count}</span>
              )
            )}
          </div>
        </div>
      );
    };

    render(<App />, document.body);
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

    document.body.querySelector('button')!.click();
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

    document.body.querySelector('button')!.click();
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

    document.body.querySelector('button')!.click();
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('$mapKeyd works', async () => {
    const todoBuffer: {
      text: string;
      done: boolean;
    }[] = [];

    const generateTodo = (listKey: number[]) => {
      return listKey.map(key => {
        if (!todoBuffer[key]) {
          todoBuffer[key] = {
            text: 'new todo ' + key,
            done: false,
          };
        }
        return todoBuffer[key];
      });
    };

    const todoList = reactive(generateTodo([1]));

    const App = () => {
      return (
        <ul>
          {$map(
            todoList,
            (item, index) => (
              <li>
                Text {item.text}
                [ID {index}]
                <button
                  onClick={() =>
                    (document.body.innerHTML =
                      document.body.innerHTML + index())
                  }
                ></button>
              </li>
            ),
            item => {
              return item.text();
            }
          )}
        </ul>
      );
    };

    render(<App />, document.body);

    // ADD
    todoList(generateTodo([1, 2, 3, 4]));

    await waitBatch();
    document.body.querySelectorAll('button').forEach(btn => btn.click());
    expect(document.body.innerHTML).toMatchSnapshot();

    // REMOVE
    todoList(generateTodo([1, 2, 4]));

    await waitBatch();
    document.body.querySelectorAll('button').forEach(btn => btn.click());
    expect(document.body.innerHTML).toMatchSnapshot();

    // MOVE
    todoList(generateTodo([1, 4, 2]));

    await waitBatch();
    document.body.querySelectorAll('button').forEach(btn => btn.click());
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('attr can be reactive', async () => {
    const App = () => {
      const count = reactive(0);
      return (
        <ul className={count}>
          <button className={count} onClick={() => count(count() + 1)}>
            CLICK {count}
          </button>
        </ul>
      );
    };

    render(<App />, document.body);
    await waitBatch();

    expect(document.body.innerHTML).toMatchSnapshot();

    document.querySelector('button')?.click();

    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('props should be reactive', async () => {
    const Hello = component<{name: string}>(({ name }) => {
      return <div>Hello {name}</div>;
    });

    const reactiveName = reactive('name');

    const App = () => {
      return <div>
        <Hello name="test" />
        <Hello name={reactiveName} />
      </div>
    };

    render(<App />, document.body);
    await waitBatch();

    expect(document.body.innerHTML).toMatchSnapshot();

    reactiveName('newName');
    await waitBatch();
    expect(document.body.innerHTML).toMatchSnapshot();

  });
});
