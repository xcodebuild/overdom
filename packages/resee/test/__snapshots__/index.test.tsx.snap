// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic $if works 1`] = `"<div><button>Count 1</button><div>Bigger than 2<span>NO 1</span></div></div>"`;

exports[`Basic $if works 2`] = `"<div><button>Count 2</button><div>Bigger than 2<span>NO 2</span></div></div>"`;

exports[`Basic $if works 3`] = `"<div><button>Count 3</button><div>Bigger than 2<span>YES 3</span></div></div>"`;

exports[`Basic $if works 4`] = `"<div><button>Count 4</button><div>Bigger than 2<span>YES 4</span></div></div>"`;

exports[`Basic $mapKeyd works 1`] = `"<ul><li>Text new todo 1[ID 0]<button></button></li><!--empty--><li>Text new todo 2[ID 1]<button></button></li><li>Text new todo 3[ID 2]<button></button></li><li>Text new todo 4[ID 3]<button></button></li></ul>0123"`;

exports[`Basic $mapKeyd works 2`] = `"<ul><li>Text new todo 1[ID 0]<button></button></li><!--empty--><li>Text new todo 2[ID 1]<button></button></li><li>Text new todo 3[ID 2]<button></button></li><li>Text new todo 4[ID 3]<button></button></li></ul>0123"`;

exports[`Basic $mapKeyd works 3`] = `"<ul><li>Text new todo 1[ID 0]<button></button></li><!--empty--><li>Text new todo 2[ID 1]<button></button></li><li>Text new todo 3[ID 2]<button></button></li><li>Text new todo 4[ID 3]<button></button></li></ul>0123"`;

exports[`Basic attr can be reactive 1`] = `"<ul class=\\"0\\"><button class=\\"0\\">CLICK 0</button></ul>"`;

exports[`Basic attr can be reactive 2`] = `"<ul class=\\"1\\"><button class=\\"1\\">CLICK 1</button></ul>"`;

exports[`Basic can reactive count 1`] = `"<button>Count 0</button>"`;

exports[`Basic can reactive count 2`] = `"<button>Count 1</button>"`;

exports[`Basic can render mutiple items 1`] = `"<div><button>CLICK</button>Count 0</div>"`;

exports[`Basic can render mutiple items 2`] = `"<div><button>CLICK</button>Count 1</div>"`;

exports[`Basic props should be reactive 1`] = `"<div><div>Hello test</div><div>Hello name</div></div>"`;

exports[`Basic props should be reactive 2`] = `"<div><div>Hello test</div><div>Hello newName</div></div>"`;

exports[`Basic should render hello world 1`] = `"<div>Hello World</div>"`;
